### DISCUSSION TITLE
react-markdown: return promise from `transformImageUri`

### DISCUSSION THREAD

**Discussion by Jiajun-Huang** [remarkjs/remark#1093] [Q&A]
I need to transfer the image URL into the firebase storage URL, which I need an async to do the job. 
However, the transformImageUri doesn't accept the function that returns a Promise.
I don't know how can I get around that, this is what I tried.

``` javascript
transformImageUri={async (src) => {
          const url = await imageUrl(src);
          console.log(url);
          return url;
        }}
```
this is what is displayed on the src for the image
![image](https://user-images.githubusercontent.com/92061424/210047249-f1f60773-6c4d-4ed2-9ad6-81d8d9b6fad5.png)

This is the component for rendering markdown
```javascript
interface Props {
  children: string;
  imageUrl: (url: string) => Promise<string>;
}

const MarkDown = ({ children, imageUrl }: Props) => {
  console.log(typeof children); // here shows object
  return (
    <div className="markdown">
      <ReactMarkdown
        remarkPlugins={[remarkMath, [remarkGfm, { singleTilde: false }]]}
        rehypePlugins={[rehypeKatex, rehypeRaw]}
        transformImageUri={async (src) => {
          const url = await imageUrl(src);
          console.log(url);
          return url;
        }}
        components={{
          code({ node, inline, className, children, ...props }) {
            const match = /language-(\w+)/.exec(className || "");
            return !inline && match ? (
              <SyntaxHighlighter
                children={String(children).replace(/\n$/, "")}
                style={oneDark}
                language={match[1]}
                PreTag="div"
                {...props}
              />
            ) : (
              <code className={className} {...props}>
                {children}
              </code>
            );
          },
        }}
      >
        {children.toString()}
      </ReactMarkdown>
    </div>
  );
};
```
This is the part of code for an article component that need Markdown component
``` javascript 
 return (
    <article>
      <Title>{title}</Title>
      <Card color="main">
        <MarkDown imageUrl={getImageUrl.bind(null, title)}>
          {data.content}
        </MarkDown>
      </Card>
    </article>
  );
```

This is the getImageUrl function
``` javascript
export async function getImageUrl(articleTitle: string, src: string) {
  const hashedTitle = getUuid(articleTitle);
  const fileRef = ref(st, "articles/" + hashedTitle + "/" + src);
  //console.log("articles/" + hashedTitle + "/" + src);
  const url = await getDownloadURL(fileRef);
  //console.log(url);
  return url;
}

```

**Comment by wooorm** [comment]
You can’t. React doesn’t provide us with the tools to make this easy either, so it can’t be added easily.

This project is a small wrapper around other tools, as a react component, that makes some small features easy to do: if you want to do more complex things than these little helpers let you do, use the tools yourself, write a plugin, those can be async

**Reply by ChristianMurphy** [reply]
Adding on, the unified docs (https://unifiedjs.com/learn/) offer an introduction to creating plugins, which work on the markdown abstract syntax tree [(mdast)](https://github.com/syntax-tree/mdast) or the HTML abstract syntax tree [(hast)](https://github.com/syntax-tree/hast) to safely edit the structure and values in content.

If you're looking for a starting point on how to find/edit URLs in an abstract syntax tree, check out https://github.com/rehypejs/rehype-external-links, which scans a document for external links and edits the HTML attributes of the links if they point to an external website.
